src/disk_io_thread.cpp

```diff
@@ -3174,10 +3139,44 @@ namespace libtorrent
 			add_completed_jobs(completed_jobs);
 	}
 
-	void disk_io_thread::thread_fun(int thread_id, thread_type_t type
-		, boost::shared_ptr<io_service::work> w)
+	bool disk_io_thread::wait_for_job(job_queue& jobq, disk_io_thread_pool& threads
+		, std::unique_lock<std::mutex>& l)
 	{
-		DLOG("started disk thread %d\n", int(thread_id));
+		TORRENT_ASSERT(l.owns_lock());
+		while (jobq.m_queued_jobs.empty())
+		{
+			// if the number of wanted threads is decreased,
+			// we may stop this thread
+			// when we're terminating the last thread, make sure
+			// we finish up all queued jobs first
+			if (threads.should_exit()
+				&& (jobq.m_queued_jobs.empty()
+					|| threads.num_threads() > 1)
```

Comment:
since num_threads() is the upper bound. is it possible the last two threads will see max_threads() as 2 and should_exit()? (i.e. is there an interleaving that could result in both threads seeing that basically)